File: schemas/content-moderation/drizzle/response_templates.ts
// response_templates: Pre-written response messages for moderator actions.
// See README.md for full design rationale.

import {
  pgEnum,
  pgTable,
  uuid,
  text,
  boolean,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { violationCategories } from "./violation_categories";
import { users } from "../../auth-rbac/drizzle/users";

export const responseTemplateActionType = pgEnum(
  "response_template_action_type",
  [
    "approve",
    "remove",
    "warn",
    "mute",
    "ban",
    "restrict",
    "escalate",
    "label",
  ]
);

export const responseTemplateScope = pgEnum("response_template_scope", [
  "global",
  "community",
]);

export const responseTemplates = pgTable(
  "response_templates",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    name: text("name").notNull(), // Internal template name (e.g., "Spam Removal — First Offense").
    actionType: responseTemplateActionType("action_type"),
    // Which moderation action this template is for. Null = usable with any action type.
    content: text("content").notNull(),
    // Template text. May include placeholders like {{username}}, {{rule}}, {{appeal_url}}.
    violationCategoryId: uuid("violation_category_id").references(
      () => violationCategories.id,
      { onDelete: "set null" }
    ), // Suggested violation category. Set null: if category is deleted, template remains.
    scope: responseTemplateScope("scope").notNull().default("global"),
    // global = available everywhere. community = specific to one community.
    scopeId: text("scope_id"), // Community ID. Null when scope = global.
    isActive: boolean("is_active").notNull().default(true),
    createdBy: uuid("created_by")
      .notNull()
      .references(() => users.id, { onDelete: "restrict" }), // Who created this template.
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdateFn(() => new Date()),
  },
  (table) => [
    index("idx_response_templates_scope").on(table.scope, table.scopeId),
    index("idx_response_templates_action_type").on(table.actionType),
    index("idx_response_templates_violation_category").on(
      table.violationCategoryId
    ),
    index("idx_response_templates_is_active").on(table.isActive),
  ]
);

File: schemas/content-moderation/spacetimedb/response_templates.rs
// response_templates: Pre-written response messages for moderator actions.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

/// Which moderation action this template is for.
/// Null action_type = usable with any action type.
// Uses ModerationActionType from moderation_actions.rs
// type: String
#[derive(SpacetimeType, Clone)]
pub enum ResponseTemplateActionType {
    Approve,
    Remove,
    Warn,
    Mute,
    Ban,
    Restrict,
    Escalate,
    Label,
}

/// Whether the template is available globally or scoped to a community.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum ResponseTemplateScope {
    Global,
    Community,
}

#[spacetimedb::table(name = response_templates, public)]
pub struct ResponseTemplate {
    #[primary_key]
    pub id: String, // UUID

    pub name: String,

    #[index(btree)]
    pub action_type: Option<ResponseTemplateActionType>, // None = usable with any action type.

    pub content: String, // May include placeholders like {{username}}, {{rule}}, {{appeal_url}}.

    #[index(btree)]
    pub violation_category_id: Option<String>, // UUID — FK → violation_categories.id (set null on delete)

    #[index(btree)]
    pub scope: ResponseTemplateScope,

    pub scope_id: Option<String>, // Community ID. None when scope = Global.
    // Composite index: (scope, scope_id) — enforce in reducer logic

    #[index(btree)]
    pub is_active: bool,

    pub created_by: String, // UUID — FK → users.id (restrict delete)

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
